pkg/internalregistry/controllers/registry_urls_observation_controller.go

```diff
@@ -0,0 +1,170 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	informers "k8s.io/client-go/informers/core/v1"
+	listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const clusterDNSSuffix = "cluster.local"
+
+type serviceLocation struct {
+	namespace string
+	name      string
+}
+
+var serviceLocations = []serviceLocation{
+	{namespace: "default", name: "docker-registry"},
+	{namespace: "openshift-image-registry", name: "image-registry"},
+}
+
+type registryURLObservation struct {
+	services               listers.ServiceLister
+	additionalRegistryURLs []string
+	cacheSyncs             []cache.InformerSynced
+	queue                  workqueue.RateLimitingInterface
+	observeRegistryURLs    func(urls []string) error
+}
+
+func NewRegistryURLObservationController(services informers.ServiceInformer, additionalRegistryURLs []string, registryURLsObservationFunc func(urls []string) error) *registryURLObservation {
+	c := &registryURLObservation{
+		services:               services.Lister(),
+		additionalRegistryURLs: additionalRegistryURLs,
+		observeRegistryURLs:    registryURLsObservationFunc,
+		cacheSyncs:             []cache.InformerSynced{services.Informer().HasSynced},
+		queue:                  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "internal-registry-services"),
+	}
+	services.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj any) bool {
+			service := obj.(*corev1.Service)
+			for _, l := range serviceLocations {
+				if l.name == service.Name && l.namespace == service.Namespace {
+					return true
+				}
+			}
+			return false
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj any) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err == nil {
+					c.queue.Add(key)
+				}
+			},
+			UpdateFunc: func(_, obj any) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err == nil {
+					c.queue.Add(key)
+				}
+			},
+		},
+	})
+	return c
+}
+
+func (c *registryURLObservation) sync(ctx context.Context, key string) error {
+	klog.V(4).InfoS("sync", "key", key)
+	// urlsForInternalRegistry returns the dns form and the ip form of the service
+	urls := append([]string{}, c.additionalRegistryURLs...)
+	for _, location := range serviceLocations {
+		urls = append(urls, urlsForInternalRegistryService(c.services, location)...)
+	}
+	if len(urls) == 0 {
+		// not expecting this, don't clobber existing values
+		klog.V(2).InfoS("no registry urls observed")
+		return nil
+	}
```

Comment:
Why not?

Reply:
This would leave the controller in a state where it could not generate a useful pull secret. I moved this logic into the imagepullsecretcontroller where I think it makes more sense.